subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -126,14 +127,20 @@ private static boolean missesDependency(Node producer, Node consumer) {
             return false;
         }
         // Do a breadth first search for any dependency
-        ArrayDeque<Node> queue = new ArrayDeque<>();
+        Deque<Node> queue = new ArrayDeque<>();
         consumer.getHardSuccessors().forEach(queue::add);
         while (!queue.isEmpty()) {
             Node dependency = queue.removeFirst();
             if (dependency == producer) {
                 return false;
             }
-            dependency.getHardSuccessors().forEach(queue::add);
+            dependency.getHardSuccessors().forEach(node -> {
+                // Using the queue itself for deduplication should be fast enough
+                // since the queue should normally be pretty short.
+                if (!queue.contains(node)) {
+                    queue.add(node);
+                }
```

Comment:
This doesn't prevent `node` from being processed multiple times. Is `A` has `B` and `C` as successors, and `C` also has `B` as successor, IIUC `B` will be added to the queue again. I think the more traditional `Set<Node> nodesSeen = new HashSet<>()` would be a better solution here.